services/media/service/migrations/committed/000022-transactional-outbox-inbox.sql

```diff
@@ -0,0 +1,235 @@
+--! Previous: sha1:95f2cb8f12acd7bc84f7438f981092e254f74973
+--! Hash: sha1:9aa991c53d2e3e092ec649dccb9607102f4928d0
+--! Message: transactional-outbox-inbox
+
+--   ___   _   _  _____  ___   ___  __  __
+--  / _ \ | | | ||_   _|| _ ) / _ \ \ \/ /
+-- | (_) || |_| |  | |  | _ \| (_) | >  < 
+--  \___/  \___/   |_|  |___/ \___/ /_/\_\
+ 
+DROP TABLE IF EXISTS app_hidden.outbox CASCADE;
+CREATE TABLE app_hidden.outbox (
+  id uuid PRIMARY KEY,
+  aggregate_type TEXT NOT NULL,
+  aggregate_id TEXT NOT NULL,
+  message_type TEXT NOT NULL,
+  segment TEXT,
+  concurrency TEXT NOT NULL DEFAULT 'sequential',
+  payload JSONB NOT NULL,
+  metadata JSONB,
+  locked_until TIMESTAMPTZ NOT NULL DEFAULT to_timestamp(0),
+  created_at TIMESTAMPTZ NOT NULL DEFAULT clock_timestamp(),
+  processed_at TIMESTAMPTZ,
+  abandoned_at TIMESTAMPTZ,
+  started_attempts smallint NOT NULL DEFAULT 0,
+  finished_attempts smallint NOT NULL DEFAULT 0
+);
+ALTER TABLE app_hidden.outbox ADD CONSTRAINT outbox_concurrency_check
+  CHECK (concurrency IN ('sequential', 'parallel'));
+
+GRANT SELECT, INSERT, DELETE ON app_hidden.outbox TO :DATABASE_GQL_ROLE;
+GRANT UPDATE (locked_until, processed_at, abandoned_at, started_attempts, finished_attempts) ON app_hidden.outbox TO :DATABASE_GQL_ROLE;
+GRANT SELECT, INSERT, UPDATE, DELETE ON app_hidden.outbox TO :DB_OWNER;
+
+-- Create the function to get the next batch of messages from the outbox table.
+DROP FUNCTION IF EXISTS app_hidden.next_outbox_messages(integer, integer);
+CREATE OR REPLACE FUNCTION app_hidden.next_outbox_messages(
+  max_size integer, lock_ms integer)
+    RETURNS SETOF app_hidden.outbox 
+    LANGUAGE 'plpgsql'
+
+AS $BODY$
+DECLARE 
+  loop_row app_hidden.outbox%ROWTYPE;
+  message_row app_hidden.outbox%ROWTYPE;
+  ids uuid[] := '{}';
+BEGIN
+
+  IF max_size < 1 THEN
+    RAISE EXCEPTION 'The max_size for the next messages batch must be at least one.' using errcode = 'MAXNR';
+  END IF;
+
+  -- get (only) the oldest message of every segment but only return it if it is not locked
+  FOR loop_row IN
+    SELECT * FROM app_hidden.outbox m WHERE m.id in (SELECT DISTINCT ON (segment) id
+      FROM app_hidden.outbox
+      WHERE processed_at IS NULL AND abandoned_at IS NULL
+      ORDER BY segment, created_at) order by created_at
```

Comment:
Very minor, but would it make sense to change this to `order by created_at ASC` to make it more explicit?

Reply:
Omitting the "ASC" is in my experience the most often used approach - adding only DESC where needed.